src/dataframe/dataframe.jl

```diff
@@ -1508,3 +1508,63 @@ end
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, index(df)[p])
 end
+
+##############################################################################
+##
+## Expand & Complete dataframes
+##
+##############################################################################
+
+
+function expanddf(df::AbstractDataFrame, indexcols::Array{Symbol,1})
+    colnames = names(df)
+    # Check to make sure the symbols in indexcols are in the df
+    for i in indexcols
+        if i ∉ colnames
+            throw(ArgumentError(":$i is not in dataframe"))
+        end
+    end
+    dummydf = similar(df[indexcols],0)
+
+    # Create a dictionary of symbol=>vector of unique values
+    uniqueVals = Dict{Symbol, Array{Any,1}}()
+    for col in indexcols
+        uniqueVals[col] = unique(df[col])
+    end
+
+    # Get a long vector of every possible combination
+    collected = collect(flatten(product(values(uniqueVals)...)))
+    dfsize = div(length(collected),length(indexcols))
+
+    # Re-format to dataframe
+    for i in 1:length(indexcols):length(collected)
+        push!(dummydf, collected[i:(i+length(indexcols)-1)])
```

Comment:
better use `view` here for performance. But seeing this code it is probably better to first construct the columns and then make a data frame from them. Here is a sample code I have written for similar thing in the past: https://github.com/PacktPublishing/Julia-1.0-Programming-Cookbook/blob/dfbb5b6c482d74f844840712dc206dcc567631ed/Chapter04/03.%20Generating%20full%20factorial%20designs/expand.jl. It does not materialize a `DataFrame` but is pretty similar.

Reply:
I considered constructing the columns first, but since there could be any number of columns, I didn't want to have to reach for metaprogramming to implement. This seemed like the most elegant approach I could think of at the time.

Reply:
maybe do something like
```
function _helper(col, c, stride, offset)
    for i in offset:stride:length(c)
        push!(col, c[i])
    end
end
for colind in 1:length(indexcols)
    _helper(dummydf[!, colind], collected, length(indexcols), colind)
end
```
(I am writing it from my head so this is not tested)

Reply:
Gotcha, I see what you're trying to get at. I'm still new to writing performance Julia code, but am I understanding correctly that building each column as an array and joining the arrays into a DF is faster?

Reply:
Yes - I am using a function barrier technique (see https://docs.julialang.org/en/latest/manual/performance-tips/#kernel-functions-1).

Reply:
I couldn't get `dummydf[!,colind]` working (the exclamation mark syntax doesn't seem to work), so I ended up with something like this:
```
function _expandhelper(vect, stride, offset)
    assembledcol = []
    for i in offset:stride:length(vect)
        push!(assembledcol, vect[i])
    end
    return assembledcol
end
```
and
```
...
dummydf = DataFrame()
    for i in 1:length(indexcols)
        hcat(dummydf,_expandhelper(collected, length(indexcols), i))
    end
...
```
I've got something like that working, but I'm not entirely sure if it's as performant as your sketch.

Reply:
You need to rebase your PR against master to make `dummydf[!, colind]` work.

Reply:
Oh! That's why - whoops. Didn't realize that there was a new release; thanks!